### PR TITLE
[BUGFIX] comparison between timezone aware and non-aware datetimes 

### PR DESCRIPTION
**Pull request by roblim**
Please annotate your PR title to describe what the PR does, then give a brief bulleted description of your PR below. PR titles should begin with [BUGFIX], [ENHANCEMENT], [FEATURE],  [DOCS], or [MAINTENANCE]. If a new feature introduces breaking changes for the Great Expectations API or configuration files, please also add [BREAKING]. You can read about the tags in our [contributor checklist](https://docs.greatexpectations.io/en/latest/contributing/contribution_checklist.html).

Changes proposed in this pull request:
- [BUGFIX] Fixed bug that was caused by comparison between timezone aware and non-aware datetimes (switched non tz-aware datetimes to aware, use .strftime("%Y%m%dT%H%M%S.%fZ") instead of .isoformat() where appropriate for run_time, assume utc time for run_time
- [DOCS] Updated docs with info on typed run ids and validation operator results
-

After submitting your PR, CI checks will run and @tiny-tim-bot will check for your CLA signature.

For a PR with nontrivial changes, we review with both design-centric and code-centric lenses.

In a design review, we aim to ensure that the PR is consistent with our relationship to the open source community, with our software architecture and abstractions, and with our users' needs and expectations. That review often starts well before a PR, for example in github issues or slack, so please link to relevant conversations in notes below to help reviewers understand and approve your PR more quickly (e.g. `closes #123`).

Previous Design Review notes:
-
-
-

Thank you for submitting!

### CODE REVIEW

File: great_expectations/core/__init__.py [hunk 1]
```diff
@@ -329,9 +329,9 @@ def __init__(self, run_name=None, run_time=None):
                     f'Unable to parse provided run_time str ("{run_time}") to datetime. Defaulting '
                     f"run_time to current time."
                 )
-                run_time = datetime.datetime.now(datetime.timezone.utc)
+                run_time = datetime.datetime.utcnow()
 
-        self._run_time = run_time or datetime.datetime.now(datetime.timezone.utc)
+        self._run_time = run_time or datetime.datetime.utcnow()
```
Comment (jcampbell) [inline great_expectations/core/__init__.py]: This doesn't quite seem like we're assuming the run_time is in UTC, but rather later formatting it as if it is UTC whether it is or isn't. I think instead we should check at this point for non-utc tz-aware datetimes and convert them to UTC.

**Review by jcampbell — COMMENTED** [review]
Question inline

**Review by jcampbell — CHANGES_REQUESTED** [review]
See comment re: handling of tz object in RunIdentifier

**Comment by roblim** [comment]
@jcampbell - ready for re-review. Updating tests now.

**Review by jcampbell — CHANGES_REQUESTED** [review]
I don't think these changes yet address the timezone issue. Consider this example for motivation:

```
import datetime
from dateutil.parser import parse

d = parse("2020-05-28T15:09:34.123456-0400")

# This is now a tz-aware datetime. If we use the current code, it will be rendered as:
d.strftime("%Y%m%dT%H%M%S.%fZ")
# "20200528T15:09:34.123456Z", which is wrong; Z = Zulu = UTC time is 4 hours later than this

# The correct approach would be
d = d.astimezone(tz=datetime.timezone.utc).strftime("%Y%m%dT%H%M%S.%fZ")
```

So, we need to EITHER render input times naive or convert on output. Further, while I still think it's right to use UTC in stores, it may be appropriate to retain local tz info when displaying in certain circumstances. I think it's probably fine to leave that for later, but I'd need to know more about the issue that caused this to know.

In this context, I think the easier thing, since we're currently *always* storing time in UTC is to convert on input. In that case, the relevant input code should be:

```
if run_time is None:
  run_time = datetime.datetime.utcnow()
elif run_time.tzinfo is not None and run_time.tzinfo.utcoffset(run_time) is not None:
  run_time = run_time.astimezone(datetime.timezone.utc)
  run_time.tzinfo = None

# we now have a *naive* datetime in utc.
```

Again, not sure we want naive yet, need to look at the issue.

**Review by jcampbell — APPROVED** [review]
Pending discussed change.

**Pull request merged**